File: Dynamic_Programming/29_Sep_2019_Coin_change/29_Sep_2019_Coin_change_space_optimized.cpp
#include<bits/stdc++.h>
using namespace std;

int count(int s[], int m, int n){
    // table[i] will be storing the number of solutions for 
    // value i. We need n+1 rows as the table is constructed 
    // in bottom up manner using the base case (n = 0) 
    int table[n+1];

    memset(table,0,sizeof(table));

    //basecase
    table[0] = 1;

    // Pick all coins one by one and update the table[] values 
    // after the index greater than or equal to the value of the 
    // picked coin 
    for(int i = 0; i < m; i ++){
        for(int j =s[i]; j <= n; j++){
            table[j] += table[j-s[i]];
        }
    }
    return table[n];
}

//driver program to test above function
int main(){
    int n = 4;
    int arr[]= {1,2,3};
    int m = sizeof(arr)/sizeof(arr[0]);
    cout <<"count: " << count(arr,m,4);
    return 0;
}

File: Dynamic_Programming/september/29_Sep_2019_Coin_change/29_Sep_2019_Coin_change_space_optimized.py
#coin change dynamic space optimized
def count(s,m,n):
    # table[i] will be storing the number of solutions for 
    # value i. We need n+1 rows as the table is constructed 
    # in bottom up manner using the base case (n = 0) 
    # Initialize all table values as 0 
    table = [0 for i in range(n+1)]

    #base case
    table[0] = 1

    # Pick all coins one by one and update the table[] values 
    # after the index greater than or equal to the value of the 
    # picked coin 
    for i in range(0,m):
        for j in range(s[i], n+1):
            table[j] += table[j-s[i]]
    return table[n]

# Driver program to test above function 
arr = [1, 2, 3] 
m = len(arr) 
n = 4
x = count(arr, m, n) 
print("count: {}".format(x))